feat: key rotation (#92)
fastify/fastify-cookie: README.md

## Options

- `secret` (`String` | `Array` | `Object`):
  - A `String` can be passed to use as secret to sign the cookie using [`cookie-signature`](http://npm.im/cookie-signature).

>   - An `Array` can be passed if you wanna use key rotation. Read more about it in [Rotating signing secret](#rotating-secret).

Reviewer: ```suggestion
  - An `Array` can be passed if key rotation is desired. Read more about it in [Rotating signing secret](#rotating-secret).
```

>   - If you want to implement more sophisticated cookie signing mechanism, you can supply an `Object` here. Read more about it in [Custom cookie signer](#custom-cookie-signer).

Reviewer: ```suggestion
  - More sophisticated cookie signing mechanisms can be implemented by supplying an `Object`. Read more about it in [Custom cookie signer](#custom-cookie-signer).
```

[cs]: https://www.npmjs.com/package/cookie#options-1

<a id="rotating-secret"></a>
### Rotating signing secret

> Key rotation is when an encryption key is retired and replaced by a generating a new cryptographic key. To implement rotation, supply an `Array` of keys to `secret` option.

Reviewer 2: by a generating -> by generating

Note:
- Key rotation is **only** achieved by redeploying the server again with the new `secret` array.
- Iterating through all secrets is an expensive process, so the rotation list should contain as few keys as possible. Ideally, only the current key and the most recently retired key.
- Although previously signed cookies are valid even after rotation, cookies should be updated with the new key as soon as possible. See the following example for how to accomplish this.

**Example:**
```js
fastify.get('/', (req, reply) => {
  const result = reply.unsignCookie(req.cookies.myCookie)

  if (result.valid && result.renew) {
    // Setting the same cookie again, this time plugin will sign it with a new key
    reply.setCookie('myCookie', result.value, {
      domain: 'example.com', // same options as before
      path: '/',
      signed: true
    })
  }
})
```

<a id="custom-cookie-signer"></a>
### Custom cookie signer

> The `secret` option optionally accepts an object with `sign` and `unsign` functions. Useful if you want to implement custom cookie signing mechanism other than what you can achieve using [Rotating signing secret](#rotating-secret). See the example below.

Reviewer: ```suggestion
The `secret` option optionally accepts an object with `sign` and `unsign` functions. This allows for implementing a custom cookie signing mechanism. See the following example:
```

[cs]: https://www.npmjs.com/package/cookie#options-1

> <a name="rotating-secret"></a>

Reviewer: ```suggestion
<a id="rotating-secret"></a>
```

### Rotating signing secret

> Key rotation is when you retire an encryption key and replace that old key by generating a new cryptographic key. To implement rotation, you can supply an `Array` of keys to `secret` option.

Reviewer: ```suggestion
Key rotation is when an encryption key is retired and replaced by a generating a new cryptographic key. To implement rotation, supply an `Array` of keys to `secret` option.
```

**Example:**
```js
fastify.register(require('fastify-cookie'), {
  secret: [key1, key2]
})
```

> Plugin will **always** use the first key to sign cookies. On the other hand, it will iterate over the supplied array to see if any of the available keys could decode the given signed cookie. This ensures that any old signed cookies are still valid.

Reviewer: ```suggestion
The plugin will **always** use the first key (`key1`) to sign cookies. When parsing incoming cookies, it will iterate over the supplied array to see if any of the available keys are able to decode the given signed cookie. This ensures that any old signed cookies are still valid.
```

> Note: Iterating through all secrets is an expensive process, so you should try to keep the rotation list as small as possible.

Reviewer: ```suggestion
Note: Iterating through all secrets is an expensive process, it is recommended to keep the rotation list as small as possible. Also, old keys should be completely removed after sufficient time has passed to allow "in the wild" cookies to have been replaced.
```

[cs]: https://www.npmjs.com/package/cookie#options-1

<a name="rotating-secret"></a>
### Rotating signing secret

Key rotation is when you retire an encryption key and replace that old key by generating a new cryptographic key. To implement rotation, you can supply an `Array` of keys to `secret` option.

**Example:**
```js
fastify.register(require('fastify-cookie'), {
  secret: [key1, key2]
})
```

> Plugin will **always** use the first key to sign cookies. But to unsign, it will iterate over the supplied array to see if any of the available keys could decode the given signed cookie. This ensures that any old signed cookies are still valid after rotation.

Reviewer: ```suggestion
The plugin will **always** use the first key (`key1`) to sign cookies. But to unsign, it will iterate over the supplied array to see if any of the available keys are able to decode the given signed cookie. This ensures that any old signed cookies are still valid after rotation.
```

Note:
- Key rotation is **only** achieved by redeploying the server again with the new `secret` array.

> - Iterating through all secrets is an expensive process, so you should try to keep the rotation list as small as possible.

Reviewer: ```suggestion
- Iterating through all secrets is an expensive process, so the rotation list should contain as few keys as possible. Ideally, only the current key and the most recently retired key.
```

> - Although previously signed cookies are valid even after rotation, you should make sure to resign them (simply another `setCookie`) as soon as possible to ensure that the subsequent request does not have to iterate over all keys to find the correct one. See example below.

Reviewer: ```suggestion
- Although previously signed cookies are valid even after rotation, cookies should be updated with the new key as soon as possible. See the following example for how to accomplish this.
```

**Example:**
```js
fastify.get('/', (req, reply) => {
  const result = reply.unsignCookie(req.cookies.myCookie)

  if (result.valid && result.renew) {
    // Setting the same cookie again, this time plugin will sign it with a new key
    reply.setCookie('myCookie', result.value, {
      domain: 'example.com', // same options as before
      path: '/',
      signed: true
    })
  }
})
```

> <a name="custom-cookie-signer"></a>

Reviewer: ```suggestion
<a id="custom-cookie-signer"></a>
```

    },
    unsign: (value) => {
      // unsign using custom logic
      return {

>         valid: true,

Reviewer 3: ```suggestion
        valid: true, // the cookie has been unsigned successfully
```

>         renew: false,

Reviewer 3: ```suggestion
        renew: false, // the cookie has been unsigned with an old secret
```
